INSP: Add path_statements warning and fix

ozkriff opened the pull request:
Based on https://github.com/rust-lang/rust/blob/1f12ac872/compiler/rustc_lint/src/unused.rs#L328-L377

Part of #5888 (_"Compiler lints support"_)

The inspection is disabled by default because of #9555.

ortem commented on src/test/kotlin/org/rust/ide/inspections/lints/RsPathStatementsInspectionTest.kt:
```diff
@@ -0,0 +1,69 @@
+import org.rust.ide.inspections.RsInspectionsTestBase
+
+class RsPathStatementsInspectionTest : RsInspectionsTestBase(RsPathStatementsInspection::class) {
+    fun `test path statement with no effect`() = checkByText("""
```
Note, `checkByText` does not check weak warnings by default:
```Kotlin
protected fun checkByText(
    @Language("Rust") text: String,
    checkWarn: Boolean = true,
    checkInfo: Boolean = false,
    checkWeakWarn: Boolean = false,
    ignoreExtraHighlighting: Boolean = false,
) = annotationFixture.checkByText(text, checkWarn, checkInfo, checkWeakWarn, ignoreExtraHighlighting)
```
If we delete `<weak_warning> ... </weak_warning>` from the code, the test still passes. In fact, it is difficult to detect such problems because the only way is to "break" the test and see if it really fails. IMO this API should be changed as it is not clear whether `checkByText` checks certain warnings or not.

@Undin I would either make `org.rust.ide.annotator.RsAnnotationTestBase#checkByText` private or remove the default values for `checkWarn`/`checkInfo`/`checkWeakWarn` parameters to avoid such problems in the future (and maybe even to find existing broken tests). WDYT?

ortem commented on src/test/kotlin/org/rust/ide/inspections/lints/RsPathStatementsInspectionTest.kt:
```diff
@@ -0,0 +1,69 @@
+            <weak_warning descr="Path statement drops value">a</weak_warning>;
+
+            let b = (DropMe,);
+            <weak_warning descr="Path statement drops value">b</weak_warning>;
```
After replacing `checkByText` with `checkWarnings` and adding a stub `Drop` definition (see the next comment), the inspection starts producing a wrong `"Path statement with no effect"` warning here

ortem commented on src/main/resources/inspectionDescriptions/RsPathStatements.html:
```diff
@@ -0,0 +1,8 @@
+Detects path statements with no effect. It is usually a mistake to have a statement that has no effect.
+
+Corresponds to the <a href="https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#path-statements">path_statements</a>
+rust warning.
```
```suggestion
Rust lint.
```

ortem commented on src/test/kotlin/org/rust/ide/inspections/lints/RsPathStatementsInspectionTest.kt:
```diff
@@ -0,0 +1,69 @@
+        }
+    """)
+
+    @ProjectDescriptor(WithStdlibRustProjectDescriptor::class)
```
Looks like we could use a stub `Drop` definition like this to avoid heavy `WithStdlibRustProjectDescriptor`
```Rust
#[lang = "drop"]
pub trait Drop {
    fn drop(&mut self);
}
```

ozkriff commented:
Fixed the review suggestions and moved out the Drop-relates improvements to another issue since the needs_drop re-implementation that is needed for them has a bunch of non-trivial corner cases and is significant in scope.

ozkriff commented on src/main/resources/messages/RsBundle.properties:
```diff
@@ -253,3 +253,5 @@ inspection.RedundantSemicolons.fix.name=Remove unnecessary trailing semicolons
 inspection.DoubleMustUse.description=This function has a `#[must_use]` attribute, but returns a type already marked as `#[must_use]`
 inspection.DoubleMustUse.FixRemoveMustUseAttr.name=Remove `#[must_use]` from the function
+
+inspection.PathStatementsInspection.description.no.effect=Path statement with no effect
```
Considering #9555 we may want to change this message for now to "Path statement drops value or has no effect" but since it's kinda confusing I'm not sure it's worth it.

vlad20012 commented:
I decided to disable it by default because it is incomplete and can give false-positives (described in #9555). The lint should be improved in further PRs and then it can be enabled by default.

bors r+
